src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLiteConnectionPoolService.cs

```diff
@@ -0,0 +1,154 @@
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SQLiteConnectionPoolService()
+        {
+        }
+
+        /// <summary>
+        /// Use a <see cref="ConcurrentExclusiveSchedulerPair"/> to simulate a reader-writer lock.
+        /// Read operations are performed on the <see cref="ConcurrentExclusiveSchedulerPair.ConcurrentScheduler"/>
+        /// and writes are performed on the <see cref="ConcurrentExclusiveSchedulerPair.ExclusiveScheduler"/>.
+        ///
+        /// We use this as a condition of using the in-memory shared-cache sqlite DB.  This DB
+        /// doesn't busy-wait when attempts are made to lock the tables in it, which can lead to
+        /// deadlocks.  Specifically, consider two threads doing the following:
+        ///
+        /// Thread A starts a transaction that starts as a reader, and later attempts to perform a
+        /// write. Thread B is a writer (either started that way, or started as a reader and
+        /// promoted to a writer first). B holds a RESERVED lock, waiting for readers to clear so it
+        /// can start writing. A holds a SHARED lock (it's a reader) and tries to acquire RESERVED
+        /// lock (so it can start writing).  The only way to make progress in this situation is for
+        /// one of the transactions to roll back. No amount of waiting will help, so when SQLite
+        /// detects this situation, it doesn't honor the busy timeout.
+        ///
+        /// To prevent this scenario, we control our access to the db explicitly with operations that
+        /// can concurrently read, and operations that exclusively write.
+        /// 
+        /// All code that reads or writes from the db should go through this.
+        /// </summary>
+        public ConcurrentExclusiveSchedulerPair Scheduler { get; } = new();
+
+        public void Dispose()
+        {
+            lock (_connectionPools)
+            {
+                foreach (var (_, pool) in _connectionPools)
+                    pool.Dispose();
+
+                _connectionPools.Clear();
+            }
+        }
+
+        public ReferenceCountedDisposable<SQLiteConnectionPool>? TryOpenDatabase(
+            Solution? bulkLoadSnapshot,
+            string databaseFilePath,
+            IPersistentStorageFaultInjector? faultInjector,
+            Action<Solution?, SqlConnection, CancellationToken> initializer,
+            CancellationToken cancellationToken)
+        {
+            lock (_connectionPools)
+            {
+                if (_connectionPools.TryGetValue(databaseFilePath, out var pool))
+                {
+                    return pool.TryAddReference() ?? throw ExceptionUtilities.Unreachable;
+                }
+
+                // try to get db ownership lock. if someone else already has the lock. it will throw
+                var ownershipLock = TryGetDatabaseOwnership(databaseFilePath);
```

Comment:
such an add named API then.  if it throws when someone else has the lock, why is it called Try? Consider an enum that represents the different types of failure modes for tihs API (up to you).

Reply:
actually, this comment doesn't to even be true.  if someone has hte lock, we'll likely get an exception that is captured by IOUtilities.PerformIO right?

Reply:
I believe this was moved without change from the old location

Reply:
I left this one since it was moved without changes.